app/docs/md/patterns/form-validation.md

```diff
@@ -0,0 +1,281 @@
+
+export async function get (req) {
+  if (req.session.problems) {
+  // 5. Back at the form we pull the problems and initial values off the session
+    let { problems, book, ...session } = req.session
+    return {
+      session,
+      // 6. The HTML page can get problems and initial values off the store
+      json: { problems, book }
+    }
+  }
+
+// 1. First user gets a blank form to fill out
+//  return { }
+}
+
+export async function post(req) {
+  const session = req.session
+  // 2. Validate form inputs and return problems
+  let { problems, book } = await validate.create(req)
+  if (problems) {
+    return {
+    // 3. Problems and initial values added to session
+      session: { ...session, problems, book },
+      // Used for progressive enhancement next module
+      json: { problems, book },
+      // 4. Redirects back to the form with the above session
+      location: '/books'
+    }
+  }
+
+  // If validation is successful the problems and old values are removed from the session
+  // eslint-disable-next-line no-unused-vars
+  let { problems: removedProblems, book: removed, ...newSession } = session
+  try {
+    const result = await upsertBook(book)
+    return {
+      session: newSession,
+      json: { book: result },
+      location: '/books'
+    }
+  }
+  catch (err) {
+    return {
+      session: { ...newSession, error: err.message },
+      json: { error: err.message },
+      location: '/books'
+    }
+  }
+}
+```
+
+## Add Problems to HTML
+To surface the problems in the frontend so that users can make adjustments we add the problems to the HTML page.
+
+### Client-side Validation
+As mentioned previously client side validation is the best way to intercept errors before they are ever sent to the server so we will setup the client side validation at the same time.
+For this example both a `url` and `text` input are required for the book tree so we will add the `required` attribute to those inputs so the browser will check for that before even submitting.
+
+Let's add the server side validation and client side validation to the page as follows:
+1. Problems and initial form values pulled out of the store.
+2. If the form is hidden by default (i.e. in summary/details) set it to open to show problems.
+3. Form problem messages at the beginning of the form.
+4. Error messages on each input with the custom element `error`
+5. Set the `value` attributes with the previous state
+6. Add client-side validation attributes to inputs
+
+
+Here is the example code annotated with these changes.
+
+```javascript
+// /app/pages/books.mjs
+
+export default function books({ html, state }) {
+  const { store } = state
+  // 1. Get Problems and values from the store
+  const book = store.book || {}
+  const problems = store.problems || {}
+
+  return html`<enhance-page-container>
+  <main>
+    <h1>New book</h1>
+    <enhance-form
+  action="/books/${book.key}"
+  method="POST">
+${'' /* 3. Overall form error messages */}
```

Comment:
Wondering about the `${ '' … }` comments here — were `//` comments not working?

Reply:
Since this is inside a block of HTML I did not want the comment to be rendered as an HTML comment. So I opted for the JS comment inside the literal so it doesn't render to the markup. But I agree it looks weird. If you think I should go with HTML comments here that would be fine.

Reply:
Ohhh, clever! I hadn't thought of that. Nice!